app/handlers/gate.go

```diff
@@ -40,32 +37,22 @@ func (g gateHandler) PostOpen(c *gin.Context) {
 	ipNet := g.getIpNetFromContext(c)
 	rules := g.createRules(ipNet)
 
-	errorDetails := make(map[string]string)
-	for _, adapter := range g.adapters {
-		callResult := adapter.CreateRules(rules)
-		if !callResult.IsSuccessful() {
-			errorDetails[adapter.ToString()] = callResult.Error.Error()
-			continue
-		}
-
-		timer := time.NewTimer(time.Duration(g.defaultTimeout))
-		go func(adapter domain.Adapter, rules []domain.Rule) {
-			<-timer.C
-			_ = adapter.DeleteRules(rules)
-		}(adapter, rules)
-	}
-
-	if len(errorDetails) > 0 {
-		c.JSON(http.StatusUnprocessableEntity, gin.H{
-			"error":   "Failed applying some rules",
-			"details": errorDetails,
-		})
+	r := g.adapterDispatcher.Open(rules)
```

Comment:
maybe it makes more sense to not "hide" failures in the ValueObject but rather always return details and an nullable `err` that will identify if something went wrong.